lib/src/signing.rs

```diff
@@ -0,0 +1,230 @@
+        Self {
+            status: SigStatus::Unknown,
+            key: None,
+            display: None,
+        }
+    }
+    pub fn invalid() -> Self {
+        Self {
+            status: SigStatus::Invalid,
+            key: None,
+            display: None,
+        }
+    }
+}
+
+/// The backend for signing and verifying cryptographic signatures.
+///
+/// This allows using different signers, such as GPG or SSH, or different
+/// versions of them.
+pub trait SigningBackend: Debug + Send + Sync {
+    /// Check if the signature was created by this backend implementation.
+    ///
+    /// Should check the signature format, usually just looks at the prefix.
+    fn is_of_this_type(&self, signature: &[u8]) -> bool;
+
+    /// Create a signature for arbitrary data.
+    ///
+    /// The `key` parameter is what `jj sign` receives as key argument, or what
+    /// is configured in the `signing.key` config.
+    fn sign(&self, data: &[u8], key: Option<&str>) -> SignResult<Vec<u8>>;
+
+    /// Verify a signature. Should be reflexive with `sign`:
+    /// ```rust,ignore
+    /// verify(data, sign(data)?)?.status == SigStatus::Good
+    /// ```
+    fn verify(&self, data: &[u8], signature: &[u8]) -> SignResult<Verification>;
+}
+
+#[derive(Debug, Error)]
+#[error("Signing error: {0}")]
+pub struct SignError(#[from] Box<dyn std::error::Error + Send + Sync>);
+
+pub type SignResult<T> = Result<T, SignError>;
+
+#[derive(Debug, Error)]
+pub enum SignInitError {
+    #[error("Unknown signing backend configured: {0}")]
+    UnknownBackend(String),
+    #[error("Failed to initialize signing: {0}")]
+    BackendError(Box<dyn std::error::Error + Send + Sync>),
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
+pub enum SignBehavior {
+    /// Drop existing signatures.
+    /// This is what jj did before signing support or does now when a signing
+    /// backend is not configured.
+    #[default]
+    Drop,
+    /// Only sign commits that were authored by self and already signed,
+    /// "preserving" the signature across rewrites.
+    /// This is what jj does when a signing backend is configured.
+    Keep,
+    /// Sign/re-sign commits that were authored by self. Foreign signatures are
+    /// overwritten on authored commits and dropped on others.
+    /// This is what jj does when configured to always sign.
+    Own,
+    /// Always sign commits, regardless of who authored or signed them before.
+    /// This is what jj does on `jj sign -f` or `jj --force-sign <snapshotting
+    /// command>`.
+    Force,
+}
+
+impl SignBehavior {
+    pub fn from_config(sign_all: bool) -> Self {
+        match sign_all {
+            true => SignBehavior::Own,
+            false => SignBehavior::Keep,
+        }
+    }
+}
+
+/// Wraps low-level signing backends and adds caching, similar to `Store`.
+#[derive(Debug, Default)]
+pub struct Signer {
+    main_backend: Option<Box<dyn SigningBackend>>,
+    other_backends: Vec<Box<dyn SigningBackend>>,
```

Comment:
What's the difference between "main" and "other"? Is "main" the one used for signing, while the others are used only for verification? Please add comments, and consider also naming them `signing_backend` and `verification_backends` if that's what they are. Actually, maybe it's simpler to always include the signing backend in the list of verification backends?